File: backend-java/src/main/java/com/peopleminer/scoring/SolvedAcStrategy.java
package com.peopleminer.scoring;

import com.peopleminer.domain.entity.Candidate;
import com.peopleminer.domain.entity.SolvedAcProfile;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.Map;

@Component
public class SolvedAcStrategy implements ScoringStrategy {

    @Override
    public String getName() {
        return "solvedAc";
    }

    @Override
    public String getDescription() {
        return "Scores based on solved.ac profile: tier, rating, problem count";
    }

    @Override
    public double getDefaultWeight() {
        return 0.20;
    }

    @Override
    public boolean isApplicable(Candidate candidate) {
        return candidate.getSolvedAcProfile() != null;
    }

    @Override
    public StrategyScore calculate(Candidate candidate) {
        SolvedAcProfile profile = candidate.getSolvedAcProfile();
        if (profile == null) {
            return StrategyScore.builder()
                .value(0)
                .breakdown(Map.of())
                .build();
        }

        Map<String, Double> breakdown = new HashMap<>();

        double tierScore = scoreTier(profile.getTier());
        breakdown.put("tier", tierScore);

        double solvedScore = scoreSolvedCount(profile.getSolvedCount());
        breakdown.put("solvedCount", solvedScore);

        double ratingScore = scoreRating(profile.getRating());
        breakdown.put("rating", ratingScore);

        double streakScore = scoreStreak(profile.getMaxStreak());
        breakdown.put("maxStreak", streakScore);

        double value = Math.round(
            tierScore * 0.35 +
            solvedScore * 0.30 +
            ratingScore * 0.25 +
            streakScore * 0.10
        );

        return StrategyScore.builder()
            .value(Math.min(value, 100))
            .breakdown(breakdown)
            .build();
    }

    private double scoreTier(int tier) {
        // Tier ranges: 0=Unrated, 1-5=Bronze, 6-10=Silver, 11-15=Gold,
        // 16-20=Platinum, 21-25=Diamond, 26-30=Ruby, 31=Master
        if (tier >= 31) return 100;  // Master
        if (tier >= 26) return 95;   // Ruby
        if (tier >= 21) return 85;   // Diamond
        if (tier >= 16) return 75;   // Platinum
        if (tier >= 11) return 60;   // Gold
        if (tier >= 6) return 45;    // Silver
        if (tier >= 1) return 30;    // Bronze
        return 10;  // Unrated
    }

    private double scoreSolvedCount(int count) {
        if (count >= 2000) return 100;
        if (count >= 1500) return 90;
        if (count >= 1000) return 80;
        if (count >= 500) return 70;
        if (count >= 300) return 60;
        if (count >= 200) return 50;
        if (count >= 100) return 40;
        if (count >= 50) return 30;
        return Math.max(count / 2, 10);
    }

    private double scoreRating(int rating) {
        if (rating >= 3000) return 100;
        if (rating >= 2500) return 90;
        if (rating >= 2000) return 80;
        if (rating >= 1500) return 70;
        if (rating >= 1200) return 60;
        if (rating >= 1000) return 50;
        if (rating >= 800) return 40;
        return Math.max(rating / 20, 20);
    }

    private double scoreStreak(int maxStreak) {
        if (maxStreak >= 365) return 100;
        if (maxStreak >= 180) return 85;
        if (maxStreak >= 100) return 70;
        if (maxStreak >= 50) return 55;
        if (maxStreak >= 30) return 45;
        if (maxStreak >= 14) return 35;
        return Math.max(maxStreak * 2, 10);
    }
}

File: backend/src/scoring/strategies/solved-ac.strategy.ts
import { Injectable } from '@nestjs/common';
import {
  ScoringStrategy,
  StrategyScore,
  CandidateWithRelations,
} from './scoring-strategy.interface';

const ADVANCED_ALGORITHM_TAGS = new Set([
  'segment_tree', 'lazyprop', 'hld', 'link_cut_tree',
  'dp_tree', 'dp_bitfield', 'dp_deque', 'dp_connection_profile',
  'flow', 'mcmf', 'bipartite_matching', 'general_matching',
  'sqrt_decomposition', 'mo', 'centroid', 'centroid_decomposition',
  'fft', 'ntt', 'convex_hull_trick', 'aliens',
  'suffix_array', 'aho_corasick', 'manacher', 'z',
  'scc', 'bcc', '2_sat', 'offline_queries',
  'persistent_ds', 'merge_sort_tree', 'rope',
]);

const INTERMEDIATE_ALGORITHM_TAGS = new Set([
  'dp', 'greedy', 'graph_traversal', 'shortest_path', 'dijkstra',
  'bfs', 'dfs', 'binary_search', 'parametric_search', 'two_pointer',
  'prefix_sum', 'stack', 'queue', 'priority_queue', 'deque',
  'trees', 'disjoint_set', 'topological_sorting', 'mst',
  'backtracking', 'divide_and_conquer', 'game_theory', 'sprague_grundy',
  'geometry', 'convex_hull', 'line_intersection', 'sweeping',
  'hash', 'trie', 'kmp', 'rabin_karp',
]);

const TIER_THRESHOLDS = {
  RUBY_I: 26,
  DIAMOND_V: 21,
  PLATINUM_V: 16,
  GOLD_V: 11,
  SILVER_V: 6,
  BRONZE_V: 1,
};

@Injectable()
export class SolvedAcStrategy implements ScoringStrategy {
  readonly name = 'solvedAc';
  readonly description = 'Scores based on solved.ac tier and algorithm depth';
  readonly defaultWeight = 0.35;

  isApplicable(candidate: CandidateWithRelations): boolean {
    return !!candidate.solvedAcProfile;
  }

  async calculate(candidate: CandidateWithRelations): Promise<StrategyScore> {
    const profile = candidate.solvedAcProfile;

    if (!profile) {
      return { value: 0, breakdown: { noProfile: 1 } };
    }

    const breakdown: Record<string, number> = {};

    breakdown.tier = this.scoreTier(profile.tier);
    breakdown.algorithmDepth = this.scoreAlgorithmDepth(profile.tagStats as Record<string, number> | null);
    breakdown.consistency = this.scoreConsistency(profile.maxStreak, profile.solvedCount);
    breakdown.classLevel = this.scoreClass(profile.classLevel, profile.classDecoration);

    const value = Math.round(
      breakdown.tier * 0.40 +
      breakdown.algorithmDepth * 0.35 +
      breakdown.consistency * 0.15 +
      breakdown.classLevel * 0.10
    );

    return {
      value: Math.min(value, 100),
      breakdown,
      metadata: {
        handle: profile.handle,
        tierName: profile.tierName,
        solvedCount: profile.solvedCount,
        rating: profile.rating,
      },
    };
  }

  private scoreTier(tier: number): number {
    if (tier >= TIER_THRESHOLDS.RUBY_I) return 100;
    if (tier >= TIER_THRESHOLDS.DIAMOND_V) return 95;
    if (tier >= TIER_THRESHOLDS.PLATINUM_V) return 85;
    if (tier >= TIER_THRESHOLDS.GOLD_V) return 70;
    if (tier >= TIER_THRESHOLDS.SILVER_V) return 55;
    if (tier >= TIER_THRESHOLDS.BRONZE_V) return 40;
    return 20;
  }

  private scoreAlgorithmDepth(tagStats: Record<string, number> | null): number {
    if (!tagStats) return 40;

    let advancedSolved = 0;
    let intermediateSolved = 0;

    for (const [tag, count] of Object.entries(tagStats)) {
      if (ADVANCED_ALGORITHM_TAGS.has(tag)) {
        advancedSolved += count;
      } else if (INTERMEDIATE_ALGORITHM_TAGS.has(tag)) {
        intermediateSolved += count;
      }
    }

    let score = 40;

    if (advancedSolved >= 100) score += 40;
    else if (advancedSolved >= 50) score += 30;
    else if (advancedSolved >= 20) score += 20;
    else if (advancedSolved >= 10) score += 10;

    if (intermediateSolved >= 200) score += 20;
    else if (intermediateSolved >= 100) score += 15;
    else if (intermediateSolved >= 50) score += 10;
    else if (intermediateSolved >= 20) score += 5;

    return Math.min(score, 100);
  }

  private scoreConsistency(maxStreak: number, solvedCount: number): number {
    let score = 40;

    if (maxStreak >= 365) score += 30;
    else if (maxStreak >= 180) score += 25;
    else if (maxStreak >= 90) score += 20;
    else if (maxStreak >= 30) score += 15;
    else if (maxStreak >= 7) score += 5;

    if (solvedCount >= 2000) score += 30;
    else if (solvedCount >= 1000) score += 25;
    else if (solvedCount >= 500) score += 20;
    else if (solvedCount >= 200) score += 15;
    else if (solvedCount >= 100) score += 10;

    return Math.min(score, 100);
  }

  private scoreClass(classLevel: number, classDecoration: string | null): number {
    let score = classLevel * 10;

    if (classDecoration === 'gold') score += 10;
    else if (classDecoration === 'silver') score += 5;

    return Math.min(score, 100);
  }
}
